packages/grafana-ui/src/components/GraphNG/GraphNG.tsx

```diff
@@ -55,310 +32,146 @@ export interface GraphNGProps {
   children?: React.ReactNode;
 }
 
-const defaultConfig: GraphFieldConfig = {
-  drawStyle: DrawStyle.Line,
-  showPoints: PointVisibility.Auto,
-  axisPlacement: AxisPlacement.Auto,
-};
-
-export const FIXED_UNIT = '__fixed';
+interface GraphNGState {
+  config: UPlotConfigBuilder;
+  data: AlignedData;
+  dataFrame: DataFrame;
+}
 
-export const GraphNG: React.FC<GraphNGProps> = ({
-  data,
-  fields,
-  children,
-  width,
-  height,
-  legend,
-  timeRange,
-  timeZone,
-  onLegendClick,
-  onSeriesColorChange,
-  ...plotProps
-}) => {
-  const theme = useTheme();
-  const hasLegend = useRef(legend && legend.displayMode !== LegendDisplayMode.Hidden);
+class UnthemedGraphNG extends React.Component<GraphNGProps, GraphNGState> {
+  constructor(props: GraphNGProps) {
+    super(props);
+    this.state = {} as GraphNGState;
+  }
 
-  const frame = useMemo(() => {
-    // Default to timeseries config
-    if (!fields) {
-      fields = {
-        x: fieldMatchers.get(FieldMatcherID.firstTimeField).get({}),
-        y: fieldMatchers.get(FieldMatcherID.numeric).get({}),
-      };
-    }
-    return outerJoinDataFrames({ frames: data, joinBy: fields.x, keep: fields.y, keepOriginIndices: true });
-  }, [data, fields]);
+  onLegendLabelClick(legend: VizLegendItem, event: React.MouseEvent) {
+    const { onLegendClick } = this.props;
+    const { fieldIndex } = legend;
 
-  const compareFrames = useCallback((a?: DataFrame | null, b?: DataFrame | null) => {
-    if (a && b) {
-      return compareDataFrameStructures(a, b);
+    if (!onLegendClick || !fieldIndex) {
+      return;
     }
-    return false;
-  }, []);
-
-  const onLabelClick = useCallback(
-    (legend: VizLegendItem, event: React.MouseEvent) => {
-      const { fieldIndex } = legend;
 
-      if (!onLegendClick || !fieldIndex) {
-        return;
-      }
-
-      onLegendClick({
-        fieldIndex,
-        mode: mapMouseEventToMode(event),
-      });
-    },
-    [onLegendClick, data]
-  );
-
-  // reference change will not trigger re-render
-  const currentTimeRange = useRef<TimeRange>(timeRange);
-
-  useLayoutEffect(() => {
-    currentTimeRange.current = timeRange;
-  }, [timeRange]);
+    onLegendClick({
+      fieldIndex,
+      mode: mapMouseEventToMode(event),
+    });
+  }
 
-  const configRev = useRevision(frame, compareFrames);
+  componentDidMount() {
+    const { timeRange, timeZone, theme } = this.props;
 
-  const configBuilder = useMemo(() => {
-    const builder = new UPlotConfigBuilder();
+    const frame = this.preparePlotFrame();
 
     if (!frame) {
-      return builder;
+      return;
     }
+    this.setState({
+      config: preparePlotConfigBuilder(frame, timeRange, timeZone, theme),
```

Comment:
we need some way to have the functions inside the confg be able to refer to the changing props/state.  I assume that means `preparePlotConfigBuilder` should be a method and not an external util function.

:thinking: -- functions that refer to `this.state.frame....` will be valid without additional caching and reinitalization -- not sure the best pattern to make that look normal

Reply:
Left it as a util function still, passing references to prop/state getters. Want to have this one testable without React ceremony.